### ISSUE TITLE
Vectorized tensor-product Clenshaw summation

### ISSUE THREAD

**Issue by maltezfaria** [IntegralEquations/IFGF.jl#27]
A large portion of the (serial) execution time of the `IFGF` forward map is spent on the
evaluation of low-order Chebyshev polynomials. Although each Chebyshev
polynomial is evaluated at multiple points, the way the `IFGF` algorithms are
built is so that we only have access to one interpolation point at a time. This
means that the *standard* vectorization which evaluates a single clenshaw
summation on a set of many points will not work for us.

An idea is to vectorize instead the evaluation of a *single point evaluation*
by exploring the tensor product structure. Essentially, we do the partial sums
by vectorizing on the coefficients, with last of the partial sums being
evaluated in non-vectorized way by the regular Clenshaw algorithm.

**Comment by maltezfaria** [comment]
For the record, vectorizing the Chebyshev polynomial evaluation by passing multiple points at a time is (probably) easier to implement. 

Some preliminary tests give a good performance improvement on the `chebeval` function, but some work is needed to organize the computations in the main algorithm so that we can indeed group target and interpolation points that are evaluated by the same Chebyshev polynomial. At present, we simply loop over points, find which cone it lies on, and evaluate the polynomial for that cone. We probably need to collect target points on the same cone in an efficient (nonallocating?) way...

**Comment by maltezfaria** [comment]
Here is a simple example that evaluates a polynomial for various points using vectorised instructions:

```Julia
using SIMD
using StaticArrays
using IFGF

# type piracy needed for the moment to use "generic" scalar code
Base.convert(::Type{Vec{N,T}},x::T) where {N,T} = Vec{N,T}(x)

x = y = z = Vec(0.,1.,2.,3.)
pt = SVector(x,y,z) # 4 points at a time (0,0,0), (1,1,1), (2,2,2) and (3,3,3)
sz = (4,4,4)
rec = IFGF.HyperRectangle((0,0,0),(1,1.,1))
c = rand(sz...)
v = IFGF.chebeval_novec(c,pt,rec,Val(sz)) # name should change
@btime IFGF.chebeval_novec($c,$pt,$rec,$(Val(sz))) # essentially same time as for a single point
```

**Comment by maltezfaria** [comment]
Implemented, with some success, in #37. We can still probably do better, but seems good enough for now.
